File: provider/src/main/java/com/raygun/raygun4android/messages/rum/RaygunRUMDataMessage.java
package com.raygun.raygun4android.messages.rum;

import com.raygun.raygun4android.messages.shared.RaygunUserInfo;

@SuppressWarnings("FieldCanBeLocal")
public class RaygunRUMDataMessage {
    private String sessionId;
    private String timestamp;
    private String type;
    private RaygunUserInfo user;
    private String version;
    private String os;
    private String osVersion;
    private String platform;
    private String data;

    public static class Builder {

        private String sessionId;
        private String timestamp;
        private String type;
        private RaygunUserInfo user;
        private String version;
        private String os;
        private String osVersion;
        private String platform;
        private String data;

        public Builder(String type) {
            this.type = type;
        }

        public Builder sessionId(String sessionId) {
            this.sessionId = sessionId;
            return this;
        }

        public Builder timestamp(String timestamp) {
            this.timestamp = timestamp;
            return this;
        }

        public Builder user(RaygunUserInfo user) {
            this.user = user;
            return this;
        }

        public Builder version(String version) {
            this.version = version;
            return this;
        }

        public Builder os(String os) {
            this.os = os;
            return this;
        }

        public Builder osVersion(String osVersion) {
            this.osVersion = osVersion;
            return this;
        }

        public Builder platform(String platform) {
            this.platform = platform;
            return this;
        }

        public Builder data(String data) {
            this.data = data;
            return this;
        }

        public RaygunRUMDataMessage build() {
            return new RaygunRUMDataMessage(this);
        }
    }

    private RaygunRUMDataMessage(Builder builder) {
        type = builder.type;
        sessionId = builder.sessionId;
        timestamp = builder.timestamp;
        user = builder.user;
        version = builder.version;
        os = builder.os;
        osVersion = builder.osVersion;
        platform = builder.platform;
        data = builder.data;
    }
}

File: provider/src/main/java/com/raygun/raygun4android/messages/rum/RaygunRUMDataMessage.kt
package com.raygun.raygun4android.messages.rum

import com.raygun.raygun4android.messages.shared.RaygunUserInfo

data class RaygunRUMDataMessage(
    val sessionId: String?,
    val timestamp: String?,
    val type: String,
    val user: RaygunUserInfo?,
    val version: String?,
    val os: String?,
    val osVersion: String?,
    val platform: String?,
    val data: String?,
) {

    class Builder(internal val type: String) {
        private var sessionId: String? = null
        private var timestamp: String? = null
        private var user: RaygunUserInfo? = null
        private var version: String? = null
        private var os: String? = null
        private var osVersion: String? = null
        private var platform: String? = null
        private var data: String? = null

        fun sessionId(sessionId: String?): Builder {
            this.sessionId = sessionId
            return this
        }

        fun timestamp(timestamp: String?): Builder {
            this.timestamp = timestamp
            return this
        }

        fun user(user: RaygunUserInfo?): Builder {
            this.user = user
            return this
        }

        fun version(version: String?): Builder {
            this.version = version
            return this
        }

        fun os(os: String?): Builder {
            this.os = os
            return this
        }

        fun osVersion(osVersion: String?): Builder {
            this.osVersion = osVersion
            return this
        }

        fun platform(platform: String?): Builder {
            this.platform = platform
            return this
        }

        fun data(data: String?): Builder {
            this.data = data
            return this
        }

        fun build(): RaygunRUMDataMessage {
            return RaygunRUMDataMessage(
                sessionId = sessionId,
                timestamp = timestamp,
                type = type,
                user = user,
                version = version,
                os = os,
                osVersion = osVersion,
                platform = platform,
                data = data,
            )
        }
    }
}
